### PR TITLE
expression, executor: rewrite built-in func makeDate using new expression evaluation architecture

### PR DESCRIPTION
**Pull request by dawxy**
Rewrite built-in func `makeDate` using new expression evaluation architecture

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -2023,35 +2023,36 @@ type makeDateFunctionClass struct {
 }
 
 func (c *makeDateFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinMakeDateSig{newBaseBuiltinFunc(args, ctx)}
+	tp := types.NewFieldType(mysql.TypeDate)
+	tp.Flen = 10
```
Comment (dawxy) [inline expression/builtin_time.go]: No field length?
[https://github.com/mysql/mysql-server/blob/5.7/sql/item_timefunc.h#L1581](url)

Comment (XuHuaiyu) [inline expression/builtin_time.go]: @dawxy 
you may refer to [this](https://github.com/mysql/mysql-server/blob/5.7/sql/item_timefunc.h#L674)

[Clion](https://www.jetbrains.com/clion/) can be used to check MySQL source code.

Comment (dawxy) [inline expression/builtin_time.go]: @XuHuaiyu 
Thank you, I found it,and the MAX_DATE_WIDTH	is 10.
refer [https://github.com/mysql/mysql-server/blob/71f48ab393bce80a59e5a2e498cd1f46f6b43f9a/sql/sql_const.h#L59](https://github.com/mysql/mysql-server/blob/71f48ab393bce80a59e5a2e498cd1f46f6b43f9a/sql/sql_const.h#L59)

**Comment by dawxy** [comment]
@hanfei1991 @XuHuaiyu , PTAL, thanks.

File: expression/builtin_time.go (hunk 1 above)
Comment (XuHuaiyu) [inline expression/builtin_time.go]: You may define MAX_DATE_WIDTH as a constant in mysql/const.go as MySQL does

File: expression/builtin_time.go [hunk 2]
```diff
@@ -2023,35 +2023,36 @@ type makeDateFunctionClass struct {
...
+// evalTime evaluates a builtinMakeDateSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_makedate
-func (b *builtinMakeDateSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinMakeDateSig) evalTime(row []types.Datum) (d types.Time, isNull bool, err error) {
+	args := b.getArgs()
+	var year, dayOfYear int64
+	year, isNull, err = args[0].EvalInt(row, b.ctx.GetSessionVars().StmtCtx)
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: sc := b.ctx.GetSessionVars().StmtCtx
so we can use it at line 2050

File: expression/builtin_time_test.go [hunk 3]
```diff
@@ -1203,42 +1206,56 @@ func (s *testEvaluatorSuite) TestTimestamp(c *C) {
...
-	Dtbl := tblToDtbl(tbl)
-	maketime := funcs[ast.MakeDate]
-	for idx, t := range Dtbl {
-		f, err := maketime.getFunction(datumsToConstants(t["Args"]), s.ctx)
-		c.Assert(err, IsNil)
-		got, err := f.eval(nil)
-		c.Assert(err, IsNil)
-		if t["Want"][0].Kind() == types.KindNull {
-			c.Assert(got.Kind(), Equals, types.KindNull, Commentf("[%v] - args:%v", idx, t["Args"]))
+		{[]interface{}{71, 1}, "1971-01-01", false, false},
```
Comment (XuHuaiyu) [inline expression/builtin_time_test.go]: add test cases for float64 value

File: mysql/const.go [hunk 4]
```diff
@@ -193,6 +193,11 @@ const (
 // AllPrivMask is the mask for PrivilegeType with all bits set to 1.
 const AllPrivMask = AllPriv - 1
 
+const (
```
Comment (XuHuaiyu) [inline mysql/const.go]: add comments for these const.

**Comment by winoros** [comment]
You can run `make dev` to make sure that you can pass CI.
And `make check` to format your code.

File: mysql/const.go [hunk 5]
```diff
@@ -193,6 +193,12 @@ const (
 // AllPrivMask is the mask for PrivilegeType with all bits set to 1.
 const AllPrivMask = AllPriv - 1
 
+// MYSQL type maximum length.
```
Comment (XuHuaiyu) [inline mysql/const.go]: s/MYSQL/MySQL

**Comment by dawxy** [comment]
@XuHuaiyu 
fixed

**Comment by dawxy** [comment]
@winoros 
Thank you, I am in the company, the system is windows, nor installed virtual machine, I now use travis-ci to test

File: expression/builtin_time.go [hunk 6]
```diff
@@ -2023,35 +2023,37 @@ type makeDateFunctionClass struct {
...
+func (b *builtinMakeDateSig) evalTime(row []types.Datum) (d types.Time, isNull bool, err error) {
+	args := b.getArgs()
 	sc := b.ctx.GetSessionVars().StmtCtx
-	year, err := args[0].ToInt64(sc)
-	if err != nil {
-		return d, errors.Trace(err)
+	var year, dayOfYear int64
+	year, isNull, err = args[0].EvalInt(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
```
Comment (hanfei1991) [inline expression/builtin_time.go]: return d, true is clearer

File: expression/builtin_time.go [hunk 7]
```diff
@@ -2023,35 +2023,37 @@ type makeDateFunctionClass struct {
...
 	}
-	dayOfYear, err := args[1].ToInt64(sc)
-	if err != nil {
-		return d, errors.Trace(err)
+	dayOfYear, isNull, err = args[1].EvalInt(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
```
Comment (hanfei1991) [inline expression/builtin_time.go]: ditto.

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**